crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -0,0 +1,818 @@
+        &self,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    );
+    fn matches_archetype(&self, archetype: &Archetype) -> bool;
+    fn matches_table(&self, table: &Table) -> bool;
+}
+
+/// A fetch that is read only. This should only be implemented for read-only fetches.
+pub unsafe trait ReadOnlyFetch {}
+
+impl WorldQuery for Entity {
+    type Fetch = EntityFetch;
+    type State = EntityState;
+}
+
+pub struct EntityFetch {
+    entities: *const Entity,
+}
+
+/// SAFE: access is read only
+unsafe impl ReadOnlyFetch for EntityFetch {}
+
+pub struct EntityState;
+
+// SAFE: no component or archetype access
+unsafe impl FetchState for EntityState {
+    fn init(_world: &mut World) -> Self {
+        Self
+    }
+
+    fn update_component_access(&self, _access: &mut FilteredAccess<ComponentId>) {}
+
+    fn update_archetype_component_access(
+        &self,
+        _archetype: &Archetype,
+        _access: &mut Access<ArchetypeComponentId>,
+    ) {
+    }
+
+    #[inline]
+    fn matches_archetype(&self, _archetype: &Archetype) -> bool {
+        true
+    }
+
+    #[inline]
+    fn matches_table(&self, _table: &Table) -> bool {
+        true
+    }
+}
+
+impl<'w> Fetch<'w> for EntityFetch {
+    type Item = Entity;
+    type State = EntityState;
+
+    #[inline]
+    fn is_dense(&self) -> bool {
+        true
+    }
+
+    unsafe fn init(_world: &World, _state: &Self::State) -> Self {
+        Self {
+            entities: std::ptr::null::<Entity>(),
+        }
+    }
+
+    #[inline]
+    unsafe fn set_archetype(
+        &mut self,
+        _state: &Self::State,
+        archetype: &Archetype,
+        _tables: &Tables,
+    ) {
+        self.entities = archetype.entities().as_ptr();
+    }
+
+    #[inline]
+    unsafe fn set_table(&mut self, _state: &Self::State, table: &Table) {
+        self.entities = table.entities().as_ptr();
+    }
+
+    #[inline]
+    unsafe fn table_fetch(&mut self, table_row: usize) -> Self::Item {
+        *self.entities.add(table_row)
+    }
+
+    #[inline]
+    unsafe fn archetype_fetch(&mut self, archetype_index: usize) -> Self::Item {
+        *self.entities.add(archetype_index)
+    }
+}
+
+impl<T: Component> WorldQuery for &T {
+    type Fetch = ReadFetch<T>;
+    type State = ReadState<T>;
+}
+
+pub struct ReadState<T> {
+    component_id: ComponentId,
+    storage_type: StorageType,
+    // NOTE: PhantomData<fn()-> T> gives this safe Send/Sync impls
+    marker: PhantomData<fn() -> T>,
+}
+
+// SAFE: component access and archetype component access are properly updated to reflect that T is read
+unsafe impl<T: Component> FetchState for ReadState<T> {
+    fn init(world: &mut World) -> Self {
+        let component_id = world.components.get_or_insert_id::<T>();
+        // SAFE: component_id exists if there is a TypeId pointing to it
+        let component_info = unsafe { world.components.get_info_unchecked(component_id) };
```

Comment:
Could these three lines be a new safe method on `world.components`?

Reply:
absolutely. very good call!